feature/experimental/p4rt/ate_tests/base_p4rt/base_p4rt_test.go

```diff
@@ -0,0 +1,414 @@
+
+	// Send ClientArbitration message on both clients.
+	clients := []*p4rt_client.P4RTClient{args.client1, args.client2}
+	for index, client := range clients {
+		if client != nil {
+			client.StreamChannelCreate(&streamlist[index])
+			if err := client.StreamChannelSendMsg(&streamlist[index].Name, &p4_v1.StreamMessageRequest{
+				Update: &p4_v1.StreamMessageRequest_Arbitration{
+					Arbitration: &p4_v1.MasterArbitrationUpdate{
+						DeviceId: streamlist[index].DeviceId,
+						ElectionId: &p4_v1.Uint128{
+							High: streamlist[index].ElectionIdH,
+							Low:  streamlist[index].ElectionIdL,
+						},
+					},
+				},
+			}); err != nil {
+				return errors.New("Errors seen when sending ClientArbitration message.")
+			}
+			if _, _, arbErr := client.StreamChannelGetArbitrationResp(&streamlist[index].Name, 1); arbErr != nil {
+				return errors.New("Errors seen in ClientArbitration response.")
+			}
+		}
+	}
+
+	// Load p4info file.
+	p4Info, err := utils.P4InfoLoad(p4InfoFile)
+	if err != nil {
+		return errors.New("Errors seen when loading p4info file.")
+	}
+	setp4Info, _ := json.Marshal(p4Info)
+
+	// Send SetForwardingPipelineConfig for p4rt leader client.
+	fmt.Println("Sending SetForwardingPipelineConfig for both clients")
+	deviceid_list := []uint64{deviceId1, deviceId2}
+	for index, client := range clients {
+		if err := client.SetForwardingPipelineConfig(&p4_v1.SetForwardingPipelineConfigRequest{
+			DeviceId:   deviceid_list[index],
+			ElectionId: &p4_v1.Uint128{High: uint64(0), Low: electionId},
+			Action:     p4_v1.SetForwardingPipelineConfigRequest_VERIFY_AND_COMMIT,
+			Config: &p4_v1.ForwardingPipelineConfig{
+				P4Info: &p4Info,
+				Cookie: &p4_v1.ForwardingPipelineConfig_Cookie{
+					Cookie: 159,
+				},
+			},
+		}); err != nil {
+			return errors.New("Errors seen when sending SetForwardingPipelineConfig.")
+		}
+	}
+
+	// Receive GetForwardingPipelineConfig
+	for index, client := range clients {
+		resp, err := client.GetForwardingPipelineConfig(&p4_v1.GetForwardingPipelineConfigRequest{
+			DeviceId:     deviceid_list[index],
+			ResponseType: p4_v1.GetForwardingPipelineConfigRequest_P4INFO_AND_COOKIE,
+		})
+		if err != nil {
+			return errors.New("Errors seen when sending SetForwardingPipelineConfig.")
+		}
+		// Compare P4Info from GetForwardingPipelineConfig and SetForwardingPipelineConfig
+		getp4Info, _ := json.Marshal(resp.Config.P4Info)
+		if string(setp4Info) == string(getp4Info) {
+			fmt.Println("P4info matches fine for client", index)
+		} else {
+			err := fmt.Errorf("P4info does not match for client %d", index)
+			fmt.Println(err.Error())
+		}
+	}
+	return nil
+}
+
+// Function to compare and check if the expected table is present in RPC ReadResponse
+func verifyReadReceiveMatch(expected_update []*p4_v1.Update, received_entry *p4_v1.ReadResponse) error {
+
+	matches := 0
+	for i, table := range received_entry.Entities {
+		if table.Entity == expected_update[i].Entity.Entity {
```

Comment:
The read response would have all the programmed entries on the target/DUT. But there is only 1 entry in the expected_update. I think we need to change "expected_update[i]" to expected_update[0] or use *p4_v1.Update as expected_update. 
Also I think we can change the comparison to use go.cmp. that would help provide flexibility to ignore unrelated fields.

Reply:
Thanks @mingyangcisco , i will take a look.

Reply:
hi @mingyangcisco , i have modified the "expected_update[i]" and also used go.cmp for Entity comparison and p4Info comparison.